docs/azure/sdk/unit-testing-mocking.md

```diff
@@ -231,6 +231,50 @@ var finder = new AboutToExpireSecretFinder(TimeSpan.FromDays(2), secretClient);
+
+:::code language="csharp" source="snippets/unit-testing/ResourceManager/ResourceManagerCodeStructure.cs" id="ParentOfVMIsRG" :::
+
+To make sure these extension methods are able to be mocked or unit tested, these methods are always implemented in this way:
+
+:::code language="csharp" source="snippets/unit-testing/ResourceManager/ComputeExtensions.cs" id="HowExtensionMethodsAreImplemented" :::
+
+All the extension methods in Azure Resource Manager SDKs are implemented in the same pattern:
+
+1. There is a class holding the actual implementation of the method, we call it the "Mocking Extension" class. It is named in this pattern "{RP Name}{Resource Name}MockingExtension". In the above example, it is `ComputeResourceGroupMockingExtension`.
+1. The extension method has a private method constructing an instance of the "Mocking Extension" class by calling its `GetCachedClient` method.
+
+Both of the methods above are instance methods, therefore they are mockable and unit-testable. In this extension methods scenario, to create a test service client, you need to create two of them, one for the "Mocking Extension", and the other is the client that this extension method is extending.
+
+## [Non-library](#tab/csharp)
+
+:::code language="csharp" source="snippets/unit-testing/ResourceManager/NonLibrary/MockComputeResourceGroupMockingExtension.cs" :::
+
+## [Moq](#tab/moq)
+
+:::code language="csharp" source="snippets/unit-testing/ResourceManager/Moq/MockComputeResourceGroupMockingExtension_TestSnippets_Moq.cs" id="Moq_GetVirtualMachines" :::
+
+# [NSubstitute](#tab/nsubstitute)
+
+:::code language="csharp" source="snippets/unit-testing/ResourceManager/NSubstitute/MockComputeResourceGroupMockingExtension_TestSnippets_NSubstitute.cs" id="NSubstitute_GetVirtualMachines" :::
+
+---
+
+There is a universal naming pattern across all the Azure Resource Manager SDKs for the "mocking extension" type:
+
+1. Find the extended type of the extension method, in the above example, it is `ResourceGroupResource`.
+2. Trim the `Resource` suffix off it and get the "Resource Name", in the above example, it becomes `ResourceGroup`.
+3. Find the namespace of the extension method, in the above example, it is `Azure.ResourceManager.Compute`.
+4. Remove the `Azure.ResourceManager` from the namespace above, and concatenate the left segments together to get the "RP Name", in the above example, it is `Compute`.
+5. Now we get the name of the mocking extension type using this pattern: `{RP Name}{Resource Name}MockingExtension`. In the above extension, the name of the mocking extension type is: `ComputeResourceGroupMockingExtension`.
+6. The mocking extension type is always in the `Mocking` sub-namespace of the extension method. In the above example, the mocking extension type is in the `Azure.ResourceManager.Compute.Mocking` namespace.
```

Comment:
well unfortunately yes, for our current solution of enabling the mocking for extension method, our customers have to know some of the internal implementation, which is also the reason I included a part of "how the extension methods are implemented".

I will try to break these steps into pieces while one of them is explaining the name pattern of the "mocking extension" and the other explaining the full process of switching the behavior of an extension method

Reply:
Sorry I am kind of occupied by something else
Here are some updates:
we are planning to write useful information on the xml document of the method itself so that our customer no longer need to know a name pattern here to get the type of mocking extension - they will find it on the xml doc of the extension method.